### PR TITLE
feat: add DocSearch with override css

### PR DESCRIPTION
**Pull request by 0618**
#### Description of changes

Add search feature using [Algolia DocSearch](https://docsearch.algolia.com/) and override the styles.

**Light**
image

image

**Dark**
image

image

#### Description of how you validated changes

Testing branch:
https://docsearch-override-css.d59dwouzxt9vx.amplifyapp.com/

Note:
there's a console error, but not introduced by this PR. It's already in `main`:

> Refused to load the image 'data:image/webp;base64,UklGRh4AAABXRUJQVlA4TBEAAAAvAQAAAAfQ//73v/+BiOh/AAA=' because it violates the following Content Security Policy directive: "img-src 'self' cm.everesttech.net amazonwebservices.d2.sc.omtrdc.net dpm.demdex.net"

https://security.stackexchange.com/questions/94993/is-including-the-data-scheme-in-your-content-security-policy-safe/
But I didn't find  a good way to resolve it.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are updated
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1202185038649077

### CODE REVIEW

File: docs/.env.example [hunk 1]
```diff
@@ -0,0 +1,3 @@
+DOCSEARCH_DOCS_APP_ID=
+DOCSEARCH_DOCS_API_KEY=
+DOCSEARCH_DOCS_INDEX_NAME=
```
Comment (wlee221) [inline docs/.env.example:3]: For my understanding, how is this filled out in prod?

Comment (0618) [reply docs/.env.example:3]: Good question! It's manually set in the Amplify app

File: docs/src/components/Layout/Header.tsx [hunk 2]
```diff
@@ -79,6 +84,11 @@ const Nav = (props) => (
 
 const Settings = ({ platform, setColorMode, colorMode }) => (
   <Flex className="docs-settings" justifyContent="center" alignItems="center">
+    <DocSearch
+      appId={process.env.DOCSEARCH_DOCS_APP_ID}
```
Comment (zchenwei) [inline docs/src/components/Layout/Header.tsx:88]: When we have our docs site switch back to SSR? I was under the impression that it is problematic with redirect back then so I switch it back to static site

**Review by zchenwei — COMMENTED** [review]
Looks great! I am not gonna approve it until you have all the issues fixed.

File: docs/src/components/Layout/Header.tsx (hunk 2 above)
Comment (0618) [reply docs/src/components/Layout/Header.tsx:88]: Did you mean the `process.env` was not available for SSR? 

I'm not sure if and when we'll switch to SSR

Comment (zchenwei) [reply docs/src/components/Layout/Header.tsx:88]: `prcoess.env` only available in SSR

Comment (reesscot) [reply docs/src/components/Layout/Header.tsx:88]: Even if this works in SSR mode, I don't understand how these `process.env` variables will be available when React hydrates this code.

Docs seem to indicate you have to make them public: https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser

Do you know how this is working?

Comment (zchenwei) [reply docs/src/components/Layout/Header.tsx:88]: Yeah, but I am not super confident that I have the entire context. It seems weird to reference `process.env` in component code though

File: docs/src/pages/_document.page.tsx [hunk 3]
```diff
@@ -26,7 +26,7 @@ const getCSPContent = (context: Readonly<HtmlProps>) => {
       font-src 'self' data:;
       frame-src 'self' *.codesandbox.io;
       img-src 'self' cm.everesttech.net amazonwebservices.d2.sc.omtrdc.net dpm.demdex.net;
-      connect-src 'self' *.shortbread.aws.dev amazonwebservices.d2.sc.omtrdc.net dpm.demdex.net ;
+      connect-src 'self' *.shortbread.aws.dev amazonwebservices.d2.sc.omtrdc.net dpm.demdex.net https://*.algolia.net https://*.algolianet.com;
```
Comment (dbanksdesign) [inline docs/src/pages/_document.page.tsx:29]: Do we need 'https://' here? I don't see it in other domains in CSP

File: docs/src/styles/docs/algolia.scss [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    border-radius: var(--amplify-components-button-border-radius);
+    color: var(--amplify-components-button-color);
+    font-size: var(--amplify-components-button-font-size);
+    line-height: var(--amplify-components-button-line-height);
+    padding-block-start: var(--amplify-components-button-padding-block-start);
+    padding-block-end: var(--amplify-components-button-padding-block-end);
+    padding-inline-start: var(--amplify-components-button-padding-inline-start);
+    padding-inline-end: var(--amplify-components-button-padding-inline-end);
+    user-select: none;
+    // all:unset removes this, so we have to add it back.
```
Comment (dbanksdesign) [inline docs/src/styles/docs/algolia.scss:23]: ahhhhh I was wondering about that.

File: docs/src/pages/_document.page.tsx (hunk 3 above)
Comment (reesscot) [reply docs/src/pages/_document.page.tsx:29]: The others probably should have the https, it's more secure to be specific with the protocol.

File: docs/src/styles/docs/algolia.scss (hunk 4 above)
Comment (0618) [inline docs/src/styles/docs/algolia.scss:23]: Try removing it

File: packages/react/src/components/AmplifyProvider/index.tsx [hunk 5]
```diff
@@ -32,14 +32,14 @@ export function AmplifyProvider({
     if (document && document.documentElement) {
       // Keep original data attributes to reset on unmount
       const originalName =
-        document.documentElement.getAttribute('data-amplify-theme');
-      const originalColorMode = document.documentElement.getAttribute(
-        'data-amplify-color-mode'
-      );
+        document.documentElement.getAttribute('data-amplify-theme') ?? name;
```
Comment (0618) [inline packages/react/src/components/AmplifyProvider/index.tsx:35]: separate it to another PR

File: docs/src/components/Layout/Header.tsx (hunk 2 above)
Comment (0618) [reply docs/src/components/Layout/Header.tsx:100]: - [ ] no new console errors
- [ ] change to use [next public expose iD ](https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser)

File: docs/src/styles/docs/algolia.scss (hunk 4 above)
Comment (0618) [reply docs/src/styles/docs/algolia.scss:23]: @reesscot  you are right. We don't need `cursor: pointer;`. Deleting

File: docs/package.json [hunk 6]
```diff
@@ -41,6 +43,7 @@
     "react-copy-to-clipboard": "^5.0.4",
     "react-dom": "17",
     "react-icons": "^4.3.1",
+    "react-instantsearch-dom": "^6.23.3",
```
[the remark this answers is not in the document]
Comment (0618) [reply docs/package.json:46]: Good catch. I forgot to remove it.

File: docs/src/components/Layout/Header.tsx (hunk 2 above)
Comment (0618) [reply docs/src/components/Layout/Header.tsx:100]: The console error was not resolved after updating the variables' name

image

Comment (0618) [reply docs/src/components/Layout/Header.tsx:100]: We actually don't need to add `NEXT_PUBLIC_` because we did it in docs/next.config.js above [ref](https://nextjs.org/docs/api-reference/next.config.js/environment-variables)

so I'll revert the change

Comment (0618) [reply docs/src/components/Layout/Header.tsx:100]: I did a quick test creating a new NextJS app and add [Docsearch](https://docsearch.algolia.com/docs/DocSearch-v3) and got the same error 😂

Comment (0618) [reply docs/src/components/Layout/Header.tsx]: Fixed the console error by adding `useEffect`

**Review by reesscot — APPROVED** [review]
LGTM! Thanks for getting search working!

File: docs/src/pages/_document.page.tsx (hunk 3 above)
Comment (dbanksdesign) [reply docs/src/pages/_document.page.tsx:29]: Ok good to know!

**Review by jacoblogan — APPROVED** [review]
LGTM, great job on the search feature!

File: docs/src/styles/docs/algolia.scss [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+    .DocSearch-Search-Icon {
+      color: inherit;
+    }
+  }
+
+  .DocSearch-Button-Key {
+    background: inherit;
+    color: inherit;
+  }
```
Comment (dbanksdesign) [inline docs/src/styles/docs/algolia.scss:46]: We'll need to add this somewhere to get the height to line up with the color mode toggle

```scss
.DocSearch-Button-Placeholder {
  font-size: var(--amplify-components-button-small-font-size);
  line-height: 1;
}
```

File: docs/src/styles/docs/algolia.scss [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    .DocSearch-Search-Icon {
+      color: inherit;
+    }
+  }
+
+  .DocSearch-Button-Key {
+    background: inherit;
+    color: inherit;
```
Comment (dbanksdesign) [inline docs/src/styles/docs/algolia.scss:45]: We'll need to add this

```scss
height: var(--amplify-font-sizes-medium);
```

File: docs/src/styles/docs/algolia.scss [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+    :focus {
+      background-color: var(--amplify-components-button-hover-background-color);
+      border-color: var(--amplify-components-button-focus-border-color);
+      color: var(--amplify-components-button-focus-color);
+      box-shadow: var(--amplify-components-button-focus-box-shadow);
+    }
+
+    .DocSearch-Search-Icon {
+      color: inherit;
```
Comment (dbanksdesign) [inline docs/src/styles/docs/algolia.scss:39]: Can we add: 

```scss
height: var(--amplify-font-sizes-medium);
width: var(--amplify-font-sizes-medium);
```

File: docs/src/styles/docs/algolia.scss [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    border-style: var(--amplify-components-button-border-style);
+    border-color: var(--amplify-colors-font-secondary);
+    border-radius: var(--amplify-components-button-border-radius);
+    color: var(--amplify-colors-font-secondary);
+    font-size: var(--amplify-components-button-font-size);
+    line-height: var(--amplify-components-button-line-height);
+    padding-block-start: var(--amplify-components-button-padding-block-start);
+    padding-block-end: var(--amplify-components-button-padding-block-end);
+    padding-inline-start: var(--amplify-components-button-padding-inline-start);
+    padding-inline-end: var(--amplify-components-button-padding-inline-end);
```
Comment (dbanksdesign) [inline docs/src/styles/docs/algolia.scss:21]:
```
    padding-block-start: var(--amplify-components-button-small-padding-block-start);
    padding-block-end: var(--amplify-components-button-small-padding-block-end);
    padding-inline-start: var(--amplify-components-button-small-padding-inline-start);
    padding-inline-end: var(--amplify-components-button-small-padding-inline-end);
```

File: docs/src/styles/docs/algolia.scss [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    box-sizing: border-box;
+    font-weight: 500;
+    transition: all var(--amplify-components-button-transition-duration);
+    background-color: inherit;
+    border-width: var(--amplify-components-button-border-width);
+    border-style: var(--amplify-components-button-border-style);
+    border-color: var(--amplify-colors-font-secondary);
+    border-radius: var(--amplify-components-button-border-radius);
+    color: var(--amplify-colors-font-secondary);
+    font-size: var(--amplify-components-button-font-size);
```
Comment (dbanksdesign) [inline docs/src/styles/docs/algolia.scss:16]:
```
    font-size: var(--amplify-components-button-small-font-size);
```

File: docs/src/components/Layout/Header.tsx [hunk 12]
```diff
@@ -33,6 +42,14 @@ export const Header = ({
...
       />
 
+      {showSearch && (
+        <DocSearch
+          appId={process.env.DOCSEARCH_DOCS_APP_ID}
+          apiKey={process.env.DOCSEARCH_DOCS_API_KEY}
+          indexName={process.env.DOCSEARCH_DOCS_INDEX_NAME}
+        />
+      )}
+
```
Comment (dbanksdesign) [inline docs/src/components/Layout/Header.tsx:52]: Right now on every page except the homepage, the search gets hidden by the sidebar. The quick fix is to move this into the `<Flex>` component on line 53.

File: docs/src/styles/docs/algolia.scss [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+  .DocSearch-Footer {
+    background: var(--amplify-colors-background-primary);
+  }
+
+  .DocSearch-Commands-Key {
+    background: inherit;
+    box-shadow: inset 0 -2px 0 0 #cdcde6, inset 0 0 1px 1px #fff,
+      0 1px 2px 1px var(--amplify-colors-shadow-primary);
+  }
```
Comment (dbanksdesign) [inline docs/src/styles/docs/algolia.scss:103]: Can we add

```
.DocSearch-Cancel {
  color: var(--amplify-components-button-link-color);
}
```

File: docs/src/styles/docs/algolia.scss [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    border-radius: var(--amplify-components-button-border-radius);
+    box-sizing: border-box;
+    font-weight: 500;
+    transition: all var(--amplify-components-button-transition-duration);
+    background-color: inherit;
+    border-width: var(--amplify-components-button-border-width);
+    border-style: var(--amplify-components-button-border-style);
+    border-color: var(--amplify-colors-font-secondary);
+    border-radius: var(--amplify-components-button-border-radius);
+    color: var(--amplify-colors-font-secondary);
```
Comment (dbanksdesign) [inline docs/src/styles/docs/algolia.scss:15]:
```
    color: var(--amplify-colors-font-tertiary);
```

**Review by dbanksdesign — COMMENTED** [review]
Just some minor styling changes. Other than that it looks good!

Before:
CleanShot 2022-05-27 at 13 27 32@2x

After:
CleanShot 2022-05-27 at 13 27 04@2x

File: docs/src/styles/docs/algolia.scss (hunk 10 above)
Comment (0618) [reply docs/src/styles/docs/algolia.scss]: Done

File: docs/src/styles/docs/algolia.scss (hunk 11 above)
Comment (0618) [reply docs/src/styles/docs/algolia.scss]: Done

File: docs/src/styles/docs/algolia.scss (hunk 13 above)
Comment (0618) [reply docs/src/styles/docs/algolia.scss:121]: Done

File: docs/src/styles/docs/algolia.scss (hunk 14 above)
Comment (0618) [reply docs/src/styles/docs/algolia.scss]: Done

**Review by dbanksdesign — APPROVED** [review]
LGTM! 

ship it
